noir_stdlib/src/hash.nr

```diff
@@ -187,6 +235,14 @@ impl<T, N> Hash for [T; N] where T: Hash {
     }
 }
 
+impl<T> Hash for [T] where T: Hash {
+    fn hash<H>(self, state: &mut H) where H: Hasher{
+        for elem in self {
+            elem.hash(state);
+        }
```

Comment:
I think we should hash the slice's length as well. What do you think?

Reply:
1) Do we want hashing slices to be distinct from hashing arrays?
2) Do we want parity with `hash_slice` in Rust?

If we added the length:
- Hashing `xs: [T; N]` would produce a distinct hash from hashing `xs.as_slice(): [T]`
- Hashing `&[]: [T]` would be distinct from hashing `()` and `[]: [T; 0]` which currently hash nothing
- I think that if `()` also hashed to something expected to be distinct, we'd have that for `x: T`, `y: U` with `x != y`, `hash(x)` would be expected to be distinct from `hash(y)`.

Comparing to `std::hash` in Rust:
- `VecDeque` and `BTreeMap` hash their lengths first
- `[T; N]`, `hash_slice`, `Vec` all appear to do a `for x in xs { hash(x) }` style loop where empty collections hash nothing
- `()` hashes nothing

I'm not sure what other benefits there are from hashing the length, but like that `hash(xs) == hash(xs.as_slice())` because it agrees with the builtin functions, e.g. `blake3_array(xs) == blake3(xs.as_slice())`.

Reply:
I was thinking for slices hasing `[ [], [] ]` should not have the same result as hashing `[ [] ]`. We don't allow nested slices anymore, but if we ever re-added them I could see this being a hard to catch issue. If we're okay with this though it may be fine. ACIR is more sensitive to each additional instruction so there's the performance argument on why we maybe should not hash the length.

I don't think we need parity with hashing in Rust in general, nor do we necessarily need hashing slices to be distinct from arrays.

Reply:
Really interesting that `VecDeque` and `Vec` have different hashing behavior in Rust, I never would have expected that.

Reply:
Ok, I've tested on my own and found that `v1` and `v2`:
```rs
let v1: Vec<Vec<u32>> = vec![ vec![] ];
let v2: Vec<Vec<u32>> = vec![ vec![], vec![] ];
```
do have distinct hashes in Rust. The comment on VecDeque::hash also indicates that hashing slices does hash the length, and that VecDeque only does it itself because its internal slices may be different lengths:
```rs
// It's not possible to use Hash::hash_slice on slices
// returned by as_slices method as their length can vary
// in otherwise identical deques.
```

Reply:
ok, sounds good: I've added the length to the hash `impl`